### PR TITLE
build: docker image can be built from sources

### PR DESCRIPTION
**Pull request by lenaschoenburg**
This adds another indirection for our dockerfile, allowing us to choose between building the image based on an already existing distball with build argument `DISTBALL=some/path` and building the distribution from sources with build argument `DIST=build`. The default remains building from a pre-existing distball but we could consider changing this in the future.

This reduces the steps to build a Zeebe image from:
```bash
mvn -DskipTests -DkskipChecks package
docker build --build-arg DISTBALL=dist/target/camunda-zeebe-*.tar.gz .
```

to
```bash
docker build --build-arg DIST=build .
```

with the final goal being:
```bash
docker build .
```

The final goal would be achieved when we switch the `DIST` argument default value from `distball` to `build`.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,29 +1,40 @@
...
     apt-get -qq update && \
     apt-get install -y --no-install-recommends tini=0.19.0-1 && \
-    cp /usr/bin/tini ${TMP_DIR}/bin/tini
+    cp /usr/bin/tini .
+COPY --link --chown=1000:0 docker/utils/startup.sh .
+
+### Build zeebe from scratch ###
+FROM maven:3-eclipse-temurin-17 as build
+WORKDIR /
+ENV MAVEN_OPTS -XX:MaxRAMPercentage=80
```
Comment (megglos) [inline Dockerfile:23]: ❓ did you run into memory issues without setting this?

File: Dockerfile [hunk 2]
```diff
@@ -1,29 +1,40 @@
...
+RUN --mount=type=cache,target=/var/apt/cache,rw \
     apt-get -qq update && \
     apt-get install -y --no-install-recommends tini=0.19.0-1 && \
-    cp /usr/bin/tini ${TMP_DIR}/bin/tini
+    cp /usr/bin/tini .
+COPY --link --chown=1000:0 docker/utils/startup.sh .
+
+### Build zeebe from scratch ###
+FROM maven:3-eclipse-temurin-17 as build
+WORKDIR /
```
Comment (megglos) [inline Dockerfile:22]: 🔧  was wondering whether we should rather have dedicated workdir (e.g. `/zeebe`), than can be assumed to be "clean", this way everything is dumped into the root of this build container?

File: Dockerfile [hunk 3]
```diff
@@ -1,29 +1,40 @@
...
+### Extract zeebe from distball ###
+FROM ubuntu:jammy as distball
+WORKDIR /
+ARG DISTBALL="dist/target/camunda-zeebe-*.tar.gz"
+COPY --link ${DISTBALL} zeebe.tar.gz
+RUN mkdir camunda-zeebe && tar xfvz zeebe.tar.gz --strip 1 -C camunda-zeebe
+
+### Image containing the zeebe distribution ###
+# hadolint ignore=DL3006
+FROM ${DIST} as dist
```
Comment (megglos) [inline Dockerfile:37]: 💭   just a nitpick, I found it a bit confusing to read personally when it states FROM DIST as dist ^^ but don't have a really better suggestion either, just adding `_SRC` at least adds some differentiation
```suggestion
FROM ${DIST_SRC} as dist
```

**Review by megglos — APPROVED** [review]
Nice quality of (dev) life improvement! Just some minor comments/suggestions/questions.

**Comment by megglos** [comment]
Just noticed, may be worth amending https://github.com/camunda/zeebe/blob/main/docs/building_docker_images.md in that regard

File: Dockerfile (hunk 3 above)
Comment (oleschoenburg) [reply Dockerfile:37]: I agree that it looks a bit confusing. But changing to `DIST_SRC` would also show up in the command line:
```
docker build --build-arg DIST_SRC=build .
```

So I'd rather keep it as short as possible

**Pull request merged**